feat:  payment methods tooltip info

frosso opened the pull request:
Fixes https://github.com/Automattic/woocommerce-payments/issues/2474

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

I am creating a separate `Tooltip` component, and using it in the checkboxes.
I considered using a separate library for this (https://github.com/Automattic/woocommerce-payments/pull/2483), but then I thought that I might get some push back. So here we are.

The reason for a separate component is that Gutenberg's `Tooltip` component doesn't play nicely when displayed within modals, causing some overflow:

![2021-07-16 17 20 24](https://user-images.githubusercontent.com/273592/126239699-b00101d7-3989-4721-95ca-93d6c0952526.gif)

In order to fix the behavior, I am using a [Portal](https://reactjs.org/docs/portals.html) to render the element in a fixed position outside of the flow.

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Ensure you have the "Enable UPE checkout" flag enabled
- Ensure you have the "Enable UPE settings changes" flag enabled
- Go to http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments
- Click on the "Add payment method" button
- Hover over the tooltip elements (or click on them ~- PLEASE NOTE: when rendered within a modal, the modal's click handler can still be a bit finnicky, especially with the icon, making the modal close - it's not a problem when rendered outside of a modal and clicking on the button~ EDIT: solved in 949bc2b1 )
- Tooltip should appear
- When clicking, clicking away can also hide the tooltip
- No more than one tooltip will be rendered at one time

![2021-07-19 15 48 14](https://user-images.githubusercontent.com/273592/126239852-fc0154b7-b394-494a-bed0-4728df3ec557.gif)

~Issue when clicking within a modal, causing the modal to close - this is happening because of the modal's own click handler and NOT because of the popover component. Hover works fine:~ EDIT: solved in 949bc2b1
![2021-07-19 21 11 59](https://user-images.githubusercontent.com/273592/126251737-7b71a4c1-2e69-43ef-a48b-0d45fc085fc3.gif)



-------------------

- [ ] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

waclawjacek commented on client/components/tooltip/tooltip-base.js:
```diff
@@ -0,0 +1,210 @@
+			setIsVisible( true );
+			return;
+		}
+
```
This is probably not an issue because likely no one is going to be providing a custom `onClose` handler (you'd have to use `TooltipBase` directly) but this function could exit prematurely here if `isVisible` is true, otherwise the timer will be set and will fire `onClose()` even when the tooltip is already invisible.

This could happen if you pass a different `hideDelay` or `onClose` handler to `useHideDelay` even if the visibility doesn't change.

waclawjacek replied:
Sorry, reading this now I realize this is not very well written.

What I had in mind was this: if the execution reaches the place where this comment is posted, it will set a timer even if the visibility hasn't changed due to the same problem as in https://github.com/Automattic/woocommerce-payments/pull/2484#discussion_r673155761.

This is resolved for `onHide` by your update to use `useRef` but not for `hideDelayMs`. I don't think this will be ran into often but if you'd like to include it, this should do the job:

```suggestion

		if ( ! isVisible ) {
			return;
		}
```

frosso replied:
Thanks for clarifying! Done 👍

waclawjacek commented on client/components/tooltip/tooltip-base.js:
```diff
@@ -0,0 +1,210 @@
+		}
+
+		// element is marked as visible, no need to hide it
+		if ( isVisibleProp ) {
```
I think this will be called if `isVisibleProp` has been true so far but you pass a different `onClose` handler or `hideDelay` value.

```suggestion
		if ( isVisibleProp && ! isVisible ) {
```

frosso replied:
Ideally the `onClose` (not `onHide`) handler _should_ maintain a stable reference.

That's a valid concern, tho - I put it in a `useRef` reference in 657dd84c , so that `onHide` can be removed from the dependencies 👍

waclawjacek reviewed (changes requested):
Fantastic PR! 🎉 The code quality is great, and all the edge cases are covered. Great work!

I am picking "Request changes" because there are some minor things that are not correct (the resize handler etc.) The other stuff are just nitpicks and do not have to be changed.

---

This is not one of the reasons why I'm selecting "Request changes" but it took me a while to process the code in `TooltipBase` because of all the `useEffect` calls with their dependency lists (trying to figure out what fires when), and things like `hasMountedRef = useRef( false )` in the custom hook. This could just be me still getting used to hooks but I felt a bit overwhelmed by all the possible scenarios I had to take into account.

I am wondering if instead of handling the `hideDelay` stuff with JS, maybe we could always mount the TooltipPortals and then use `display: none` to hide them, added at the last stage of the CSS transition? I think this would reduce the complexity a bit.

frosso commented on client/components/tooltip/tooltip-base.js:
```diff
@@ -0,0 +1,210 @@
+
+		const debouncedCalculation = debounce( calculateTooltipPosition, 150 );
+
+		document.addEventListener( 'resize', debouncedCalculation );
```
You're right, thanks, done in e1d13d3e

frosso commented on client/components/tooltip/test/tooltip.test.js:
```diff
@@ -0,0 +1,133 @@
+/** @format **/
```
Sure thing! Force of habit on my part, I'll have to be more careful in the future - done in 92f69d78

frosso commented on client/components/tooltip/index.js:
```diff
@@ -0,0 +1,51 @@
+		>
+			<TooltipBase
+				{ ...props }
+				onClose={ handleClose }
```
I like it, looking around on other libraries they also use a `hide`/`visible` vocabulary 👍 
 Done in 3cf29363

frosso commented on client/components/payment-methods-checkboxes/payment-method-checkbox.js:
```diff
@@ -3,16 +3,51 @@
+import Tooltip from '../tooltip';
 import './payment-method-checkbox.scss';
 
+const infoMap = {
```
Yeah, I think at some point it might be useful to extract things to separate consts (at least the payment method enums).
But I think it's good to colocate what is needed rather than placing everything in a "constants" file - so for this reason I'd rather keep this map here for now, since it's not needed anywhere else :)

frosso replied:
I misunderstood the comment, sorry about that! Fixed in 4d5c1cdc

frosso commented:
> This is not one of the reasons why I'm selecting "Request changes" but it took me a while to process the code in `TooltipBase` because of all the `useEffect` calls with their dependency lists (trying to figure out what fires when), and things like `hasMountedRef = useRef( false )` in the custom hook. This could just be me still getting used to hooks but I felt a bit overwhelmed by all the possible scenarios I had to take into account.

> I am wondering if instead of handling the `hideDelay` stuff with JS, maybe we could always mount the TooltipPortals and then use `display: none` to hide them, added at the last stage of the CSS transition? I think this would reduce the complexity a bit.

I could always mount the portal, but then I would also need to reset the styles.
I chose the `visibility` CSS rule because it allows us to calculate the dimensions of the hidden elements (i.e.: the tooltip content). When using `display: none`, the hidden element is removed from the document flow and gets `0` dimensions.
I need the element to be part of the document flow to calculate its width for positioning, then placing it in the fixed position. Otherwise there could a flash where the tooltip is not placed in the correct spot (or, worse, in a previous spot).
I could place it outside of the flow with `position: fixed: top: -9999px; left: -9999px;`, but then I feel like we're just trading one thing for the other 😅 
Another thing that I tried to balance when writing this is that I also tried to ensure that we can use the transition on `opacity` on mount/unmount of the element.

waclawjacek reviewed (changes requested):
Everything seems great and I really hate to click **Request changes** again but a bug seems to have been introduced in 49d7db7b9fea7f378272566248dce7a3cee1ea22:

Previously, the UI was very snappy: https://d.pr/v/2RWeS3/ms2DGYzSdN
Now the tooltips linger: https://d.pr/v/QeoNWq/YF3jOp6ldB

frosso commented:
@waclawjacek thanks for pointing that out, fixed 👍

waclawjacek reviewed (approved):
:shipit: Let's gooooooooo!
